### PR TITLE
feat(linalg): matmul + transpose backed by OpenBLAS / Accelerate / cuBLAS

### LINKED ISSUE #11: [Feature] Linear algebra: matmul + transpose (OpenBLAS / cuBLAS)
# [Feature] Linear algebra: matmul + transpose (CPU OpenBLAS / CUDA cuBLAS)

> **Type:** `[Feature]` — BLAS-backed linear algebra
> **Roadmap reference:** Eager Tensor Engine, README Phase 1 item 5
> **Blocks:** `nn::Linear`, attention, every transformer layer, autograd's `Linear` backward
> **Depends on:** Issue 02 (Tensor + dispatch), Issue 03 (parity harness)

---

## 1. Background

Element-wise ops (Issue 03), reductions (Issue 04), and indexing (Issue 05) cover roughly 60 % of the surface of a deep-learning tensor library. The remaining headline primitive is **matrix multiplication** — the op that dominates training time for any transformer-class model and the linchpin of `nn::Linear`, attention, and convolutional reductions.

We don't write our own GEMM. The plan is to **link against industry-strength BLAS** (OpenBLAS / Apple Accelerate on CPU, cuBLAS on CUDA) and present a unified front door. Hand-rolled fused matmul kernels are a Phase 4 concern; for now, the goal is functional correctness, broadcasting semantics that match PyTorch, and end-to-end coverage in the parity harness.

`transpose` is shipped in the same issue because it's the cheap zero-copy operation that any BLAS-backed code path needs to canonicalise its inputs into row-major contiguous form (or supply the `'T'` flag to GEMM directly).

---

### 2.1 Functional

- **F1.** `Tensor matmul(const Tensor& a, const Tensor& b)` matching PyTorch's `torch.matmul` rules:
  - 1-D × 1-D ⇒ scalar (dot product).
  - 2-D × 2-D ⇒ standard GEMM.
  - 1-D × 2-D ⇒ row vector × matrix (returns 1-D).
  - 2-D × 1-D ⇒ matrix × column vector (returns 1-D).
  - N-D × M-D (N, M ≥ 2) ⇒ batched matmul with broadcasting on the leading batch dims, GEMM on the trailing two.
- **F2.** Operator overload `operator@` is **not** added (C++ has no `@` operator). The free function form is the only public surface.
- **F3.** `Tensor transpose(const Tensor& x, int dim0, int dim1)` returns a zero-copy view with the two dims swapped. Negative dims are normalised. `dim0 == dim1` returns the input unchanged.
- **F4.** `Tensor::T()` (no args) shorthand: only valid for 2-D tensors, returns `transpose(*this, 0, 1)`.
- **F5.** Supported dtypes for `matmul`: `float32`, `float64`. Mixed dtypes promote per Issue 03's promotion table; integer matmul is rejected with `DTypeError` (consistent with how `div` rejects integers).
- **F6.** Output device matches the input device. Cross-device throws `DeviceError`. Both inputs must be on the same device.
- **F7.** Shape mismatch (incompatible inner dim) raises `ShapeError` with the offending shapes in the message.
- **F8.** `matmul` is registered into the dispatch table with separate CPU and CUDA backends.

### 2.2 Non-functional

- **N1.** Numerical parity vs PyTorch reference: float32 within `1e-4` rel for matrices up to 1024×1024 (BLAS routines are fused-multiply-add by default and drift slightly from the naïve reference); float64 within `1e-12`.
- **N2.** CUDA `matmul` on a 1024×1024 fp32 problem matches `cublasSgemm` to ≤ 1.05× — i.e., we use the BLAS call directly with no measurable wrapper overhead.
- **N3.** No heap allocation in the matmul hot path beyond the output tensor and a single fixed-size strided-input materialisation buffer reused per call.
- **N4.** `transpose` is O(1) in tensor numel (pure metadata mutation).
- **N5.** All linalg ops appear in `docs/ops.md` with shape rules, supported dtypes, and a note about BLAS fused-multiply-add precision.

---

## 3. Feature Description

```cpp
// In <ctorch/ops/linalg.h>
namespace ctorch {
Tensor matmul(const Tensor& a, const Tensor& b);
Tensor transpose(const Tensor& x, int dim0, int dim1);
}

class Tensor {
public:
  Tensor T() const; // 2-D shorthand for transpose(*this, 0, 1).
};
```

End-to-end:

```cpp
// nn::Linear-like usage
auto pre  = ctorch::matmul(x, weight.T()) + bias;            // (B, out_features)
auto attn = ctorch::matmul(q, ctorch::transpose(k, -1, -2)); // (B, H, T, T)
```

---

### 4.1 BLAS adapter layer

A small `src/linalg/blas.h` declares a thin device-agnostic interface:

```cpp
namespace ctorch::linalg {
void gemm_cpu_f32(bool ta, bool tb, int M, int N, int K,
                  float alpha, const float* a, int lda,
                  const float* b, int ldb, float beta, float* c, int ldc);
void gemm_cpu_f64(...);
void gemm_cuda_f32(cublasHandle_t, /* same args */);
void gemm_cuda_f64(...);
}
```

CPU side picks OpenBLAS or Accelerate based on what's found at configure time (CMake's `BLAS::BLAS` target). CUDA side uses cuBLAS via a per-device handle cached in a thread-local. We deliberately don't expose `cublasHandle_t` outside the linalg layer.

### 4.2 Front-door

`src/ops/matmul.cpp` (CPU front-door TU, mirrors `binary_ops_cpu.cpp`):

```cpp
Tensor matmul(const Tensor& a, const Tensor& b) {
  check_devices_match(a, b, "matmul");
  check_no_integer(a, b, "matmul");
  const auto promoted = promote_types(a.dtype(), b.dtype());
  Tensor a2 = maybe_cast(a, promoted);
  Tensor b2 = maybe_cast(b, promoted);
  // Decide shape based on a.ndim() / b.ndim() — see F1.
  Tensor out(out_shape, promoted, a.device());
  dispatch::call<op::MatmulOp>(a.device().kind, a2, b2, out);
  return out;
}
```

The shape-decision logic is a small switch on `(a.ndim(), b.ndim())` covering the five PyTorch cases. Helpers live in `src/ops/matmul_shape.cpp` so the CPU and CUDA backends share them.

### 4.3 CPU backend

`src/ops/matmul_cpu.cpp` materialises the inputs into row-major contiguous buffers if they aren't already, then issues one or more `cblas_sgemm`/`cblas_dgemm` calls. Batched matmul over leading dims becomes a loop of GEMM calls (a `cblas_sgemm_batch` upgrade is a follow-up perf optimisation).

The strided-to-contiguous materialisation reuses the same `for_each_n_unary` machinery from Issue 03 — no new infrastructure. The buffer is a single `std::vector<std::byte>` allocated once per call, which is the only allowed heap allocation per N3.

### 4.4 CUDA backend

`src/ops/matmul_cuda.cu` issues a `cublasSgemm`/`cublasDgemm` call inside a `cuda::DeviceGuard`. cuBLAS uses **column-major** storage by convention, but the standard trick of computing `C^T = B^T A^T` lets us feed row-major buffers directly with the right `op` flags — no extra transpose kernel needed.

For batched matmul we use `cublasSgemmStridedBatched` when the leading-dim layout is contiguous; otherwise fall back to a loop. The decision is made on the host before the launch.

### 4.5 `transpose` and `Tensor::T()`

Pure metadata, lives in `src/tensor.cpp` next to `permute` (which is the general-rank superset). `transpose(x, i, j)` is `x.permute(perm_with_i_and_j_swapped)`; `Tensor::T()` calls `transpose(*this, 0, 1)` after asserting `ndim == 2`.

### 4.6 Op keys

```cpp
struct MatmulOp {
  using fn_t = void(*)(const Tensor&, const Tensor&, Tensor&);
};
```

### 4.7 Files added by this issue

```
include/ctorch/
└── ops/
    └── linalg.h              # matmul + transpose declarations

src/
├── linalg/
│   ├── blas.h                # device-agnostic gemm thin wrapper
│   ├── blas_cpu.cpp          # OpenBLAS / Accelerate dispatch
│   └── blas_cuda.cu          # cuBLAS handle cache + gemm wrapper
├── ops/
│   ├── matmul.cpp            # public front-door + shape rules + registrar
│   ├── matmul_cpu.cpp        # CPU backend (calls into linalg/blas_cpu)
│   ├── matmul_cuda.cu        # CUDA backend (calls into linalg/blas_cuda)
│   └── matmul_shape.cpp      # 1D×1D, 1D×2D, …, batched broadcasting
└── tensor.cpp                # extend with transpose / T

cmake/
└── ctorch_blas.cmake         # find_package(BLAS) + cuBLAS toggle

tests/ops/
├── matmul_shape_test.cpp     # 1D×1D, 1D×2D, 2D×1D, 2D×2D, batched broadcasting
├── matmul_dtype_test.cpp     # f32 / f64 promotion, integer rejection
└── transpose_test.cpp        # T(), transpose(x, i, j), permute parity

tests/parity/
└── matmul_parity_test.cpp

tests/bench/
└── matmul_bench.cu           # cuBLAS vs our front door

docs/ops.md                   # extend with matmul + transpose rows
```

CMake additions:
- `cmake/ctorch_options.cmake` already toggles `CTORCH_CUDA`. Add a new `CTORCH_BLAS` option (default `ON`) so CPU-only smoke builds without OpenBLAS still compile (with a runtime "no BLAS available" error if `matmul` is invoked).

---

## 5. Acceptance Criteria

- [ ] All five `matmul` shape cases (1D×1D, 1D×2D, 2D×1D, 2D×2D, batched) pass parity tests on both CPU and CUDA.
- [ ] Mixed `float32` × `float64` promotes to `float64`; integer matmul raises `DTypeError`.
- [ ] `transpose(x, i, j)` is zero-copy: result `.storage().data()` equals `x.storage().data()`, modifying source visible in result.
- [ ] `Tensor::T()` rejects non-2D inputs with `ShapeError`.
- [ ] `matmul` cross-device input throws `DeviceError`; shape-mismatch throws `ShapeError` with both offending shapes printed.
- [ ] CUDA `matmul` benchmark on 1024×1024 fp32 lands within 5 % of a direct `cublasSgemm` call (proves the wrapper is thin).
- [ ] CPU `matmul` of two 256×256 fp32 matrices is at least 5× faster than a hand-written triple-loop reference (proves BLAS is actually wired in).

---

## 6. Suggested commit breakdown

1. `feat(tensor): transpose + T() zero-copy views`
2. `build: find_package(BLAS) + CTORCH_BLAS option + cuBLAS link`
3. `feat(linalg): blas thin wrapper for sgemm/dgemm (CPU + CUDA paths)`
4. `feat(ops): matmul shape rules + front-door + dispatch key`
5. `feat(ops): CPU matmul backend wired through OpenBLAS/Accelerate`
6. `feat(ops): CUDA matmul backend wired through cuBLAS (incl. strided-batched)`
7. `chore(test): commit matmul parity fixtures + extend gen_parity.py`
8. `bench: matmul vs cublasSgemm direct call`

---

## 7. Risks and trade-offs

| Risk | Mitigation |
| --- | --- |
| OpenBLAS not installed on user's system | CMake `find_package(BLAS REQUIRED)` with a clear error; document the install command per platform in README |
| Apple Accelerate's cblas signature differs subtly from OpenBLAS | Wrap behind `linalg/blas_cpu.cpp` so the front-door TU is indifferent |
| cuBLAS column-major vs our row-major causing transposed results | Document the `C^T = B^T A^T` trick once in `linalg/blas_cuda.cu`; lock in via a parity test that has asymmetric M, N, K |
| Batched matmul with a non-contiguous batch stride forces a loop fallback | Acceptable for correctness milestone; hot-path optimisation is a Phase 4 concern |
| `cublasSgemm` numerical drift vs naive reference | Tolerance set to `1e-4` rel for fp32 in N1; documented in `docs/ops.md` |
| Multiple cuBLAS handles per device increase setup cost | Cache one handle per device in a thread-local |

---

## 8. Out of scope

- Hand-written / fused matmul kernels (e.g., split-K, tensor cores) — Phase 4 perf milestone.
- `bmm` / `addmm` / `baddbmm` convenience wrappers — separate `[Feature]` issue if a user requests them.
- `einsum` — separate `[Feature]` issue.
- `dot` / `vdot` / `inner` / `outer` — fold into a follow-up "vector ops" issue.
- Sparse matmul, complex dtypes — not in roadmap until much later.
- Backward / autograd integration — Autograd milestone.
- Operator overload (no `@` in C++).

### PR DESCRIPTION
**Pull request by Hayden727**
## Summary
Implements [issue #11](https://github.com/Hayden727/ctorch/issues/11) — BLAS-backed matrix multiplication and zero-copy transpose.

- `Tensor::T()` and `ctorch::transpose(x, i, j)` — zero-copy view ops; T() is the 2-D shorthand and rejects non-2-D inputs with `ShapeError`.
- `ctorch::matmul` — full PyTorch shape rules (1-D × 1-D dot, 1-D × 2-D, 2-D × 1-D, 2-D × 2-D, batched N-D with right-aligned broadcast). f32/f64 only; integer / bool raise `DTypeError`; cross-device raises `DeviceError`; inner-dim mismatch raises `ShapeError` with both shapes printed.
- BLAS layer — `linalg/blas.h` thin wrapper, `blas_cpu.cpp` dispatches into `cblas_sgemm`/`dgemm` via `find_package(BLAS)` (Apple Accelerate, OpenBLAS, MKL all work). `blas_cuda.cu` uses `cublasSgemm`/`Dgemm` via a per-thread per-device handle cache, with the standard `C^T = B^T A^T` trick for cuBLAS column-major.
- CMake — new `CTORCH_BLAS` option (default ON) gated through `cmake/ctorch_blas.cmake`. Honours `BLA_VENDOR`, locates `cblas.h` in non-system prefixes (Homebrew, Apt). CUDA build links `CUDA::cublas`. `CTORCH_BLAS=OFF` is a supported smoke build (matmul becomes a stub; CPU matmul tests gated off).
- 273 mac tests pass; full dell GPU run (CUDA 12.0, sm_75/80/86) is **clean** for matmul (the only failure is a pre-existing `SumBench.CudaWithin15PercentOfCub` perf flake from issue #9, unrelated).

## Test plan
- [x] Mac local CPU build (Apple Accelerate): `ctest -j4` → 273/273 pass.
- [x] Dell remote CUDA build (`CTORCH_CUDA=ON`, `CTORCH_BLAS=OFF`): `ctest -j` → 266/267 pass; only failure is the unrelated perf flake.
- [x] New `matmul_cuda_test` (4 cases): `TwoDByTwoDMatchesHandComputed`, `AsymmetricMNKLockstheColumnMajorTrick` (asymmetric M=5/N=11/K=7 to lock in the cuBLAS row→col reinterpretation), `MatmulOfTransposeUsesCpuRoundTripFallback` (non-contig CUDA), `MixedFloatPromotesToFloat64` (mixed dtype CUDA) — all pass.
- [x] All 5 matmul shape rules + 8 parity fixtures via `np.matmul`.

Closes #11

### CODE REVIEW

**Comment by Hayden727** [comment]
@codex review

**Comment by Hayden727** [comment]
@codex review.

**Comment by Hayden727** [comment]
@codex review.

**Comment by Hayden727** [comment]
@codex review.

**Comment by Hayden727** [comment]
@codex review.
